### ISSUE TITLE
Simple Domains Like Google.com are Getting `getaddrinfo EAI_AGAIN` Errors

### ISSUE THREAD

**Issue by tucker19** [louislam/uptime-kuma#114]
**Describe the bug**
I have about 10 domains I am watching, 7 of my own domains and 3 common domains, I recently installed piHole and I noticed 3 of my domains, 2 common and 1 of mine, started failing all the time. I looked into it and seems as these all are getting `getaddrinfo EAI_AGAIN` errors. I would understand if it was only my own domains but the common ones like Google and Reddit are weird. I looked at the logs within Portainer and I am not seeing any errors being thrown only `Heartbeat <number>`. The two common domains that are failing are `www.google.com` and `www.reddit.com`. If I click on the links within Uptime-Kuma I successfully get to the sites.

**Expected behavior**
Better Logging?

**Desktop (please complete the following information):**
 - Uptime Kuma Version: 1.0.6
 - Using Docker?: Yes
 - OS: Ubuntu 20.04 LTS
 - Browser: Firefox/Chrome

**Additional context**
Uptime-Kuma is in same Docker-Compose as piHole and have house router running through piHole now.

**Comment by louislam** [comment]
This error is from Node.js, it is the best error message we can get.

I never use piHole, does it replace your DNS server on your Ubuntu machine?

"EAI_AGAIN is actually a DNS lookup timed out error means it is either a network connectivity error or some proxy related error."
https://www.codingdefined.com/2015/06/nodejs-error-errno-eaiagain.html

If piHole replaced your dns, you should try other dns like 8.8.8.8

**Comment by tucker19** [comment]
> This error is from Node.js, it is the best error message we can get.
> 
> I never use piHole, does it replace your DNS server on your Ubuntu machine?
> 
> "EAI_AGAIN is actually a DNS lookup timed out error means it is either a network connectivity error or some proxy related error."
> https://www.codingdefined.com/2015/06/nodejs-error-errno-eaiagain.html
> 
> If piHole replaced your dns, you should try other dns like 8.8.8.8

Yes, it replaces the DNS server on the Ubuntu box as I have shutdown the internal DNS server and use the DNS entries that I get from my router. Also within piHole and my router I have 8.8.8.8 and 1.1.1.1 as backups if the primary DNS is not available. Its just weird that its only some of my domains failing 2 of which are very common domains like Google.com or Reddit.com

**Comment by rezzorix** [comment]
> Yes, it replaces the DNS server on the Ubuntu box as I have shutdown the internal DNS server and use the DNS entries that I get from my router. Also within piHole and my router I have 8.8.8.8 and 1.1.1.1 as backups if the primary DNS is not available. Its just weird that its only some of my domains failing 2 of which are very common domains like Google.com or Reddit.com

From what I read it is a misconfiguration in either your pi hole and/or router and/or (docker-) network config - not with kuma.
Therefore I would not label this issue as bug in here for the time being.

I have 2 pi-holes running in my network (main + backup), all DNS traffic is forced  to get all information from these 2 only.
Kuma works without any problem, monitoring more than a handful different servers throughout all versions since it was released.

Look at your pi hole and router config first and see if the issue is there.

If you like me to have a look at your pi hole and router config please share some more details here.
What router are you using? Could you post your docker-compose.yml?

**Comment by louislam** [comment]
Yes, as I always said, Uptime Kuma is very simple. It uses axios to make a request and nothing special.

https://github.com/louislam/uptime-kuma/blob/657acf748bd6af1a3e6bfcb306500b216c953f7c/server/model/monitor.js#L74-L76

Since piHole replaced your dhcp and dns, your router dns can be skipped. I still suggest you test it without piHole.

Furthermore, you can test curl with docker on your Ubuntu machine.
```bash
docker run curlimages/curl https://google.com
```

**Comment by rezzorix** [comment]
> Since piHole replaced your dhcp and dns, your router dns can be skipped. I still suggest you test it without piHole.

piHole isnt necessarily replacing the dhcp (i for example have my router manage dhcp because of different vlans).

But as @tucker19 wrote he has DNS set in router and back-dns set in router, I suspect something in between router and piHole settings causes a hiccup when dns requests are happening.

**Comment by tucker19** [comment]
> 
> 
> Yes, as I always said, Uptime Kuma is very simple. It uses axios to make a request and nothing special.
> 
> https://github.com/louislam/uptime-kuma/blob/657acf748bd6af1a3e6bfcb306500b216c953f7c/server/model/monitor.js#L74-L76
> 
> Since piHole replaced your dhcp and dns, your router dns can be skipped. I still suggest you test it without piHole.
> 
> Furthermore, you can test curl with docker on your Ubuntu machine.
> 
> ```shell
> docker run curlimages/curl https://google.com
> ```

Thank you for this response. I actually just tested all domains in question with the docker curl and now they are working just fine. I did have to add `www.` in front of them all to get them working but that is nothing. I'll be closing this conversation.

**Comment by vitobotta** [comment]
Hi! I am having this exact issue quite frequently and am not using PiHole or anything else. Just Cloudflare's resolvers directly. This is with a VPS on Hetzner Cloud, but I have many other servers with the same provider and configuration and never had any DNS related issues. What can I try? Thanks!

**Comment by antoine2tt** [comment]
Hi,

I had the same issue. What I did is set a retry to avoid this issue !

**Comment by louislam** [comment]
@vitobotta @antoine2tt

My conclusion is still look like that it is your network's problem that you are not realized before, since Uptime Kuma is a very simple app based on Node.js and it is working for most people.

Let say if you want to host another Node.js app that required to connect to another server, you are probably facing the same error.

It could be related to DNS server, proxy server or physical network.

**Comment by vitobotta** [comment]
The weird thing is that I don't have any DNS issues with other apps, I use Cloudflare's resolvers and my network is pretty stable - I even run Kubernetes with etcd in HA mode which is quite sensitive to network issues.

**Comment by rezzorix** [comment]
Well, weird thing or not, it all points towards an issue within your specific network. My guess is that the error is somewhere in your virtual networks (that you do have if I understood u right). That would be a starting point to see if an error can be narrowed down / identified there.  As this is a very individual setup nobody else than you have to narrow down why the error occurs.

**Comment by antoine2tt** [comment]
Well just wanted to add that uptime Kuma is running on a dedicated hosted server (which I pay for) 
It's strange that a server hoster can have that type of network issue..

**Comment by rezzorix** [comment]
> 
> 
> Well just wanted to add that uptime Kuma is running on a dedicated hosted server (which I pay for)
> It's strange that a server hoster can have that type of network issue..

Your server hoster is not responsible for your servers network setup. 
Especially not on a dedicated server.

You most likely have several (virtual) network interfaces on your server; each active / inactive, with / without (external/internal) IP addresses assigned.

I say chances are high that there are hiccups on your servers end rather than with uptime-kuma 😉

**Comment by antoine2tt** [comment]
Yes sure ! It could be network related. 

Anyway, with the "retries" I setup, I can avoid theses kind of errors 👌

**Comment by neldonado** [comment]
I am running into the same type of errors, I am running uptime kuma off a google cloud server running an ubuntu 20.04 LTS VM

It'll say UP and then bounces back with that error. Back and forth back and forth.

**Comment by rezzorix** [comment]
> I am running into the same type of errors, I am running uptime kuma off a google cloud server running an ubuntu 20.04 LTS VM
> 
> It'll say UP and then bounces back with that error. Back and forth back and forth.

Can you be more specific?
You running it on a cloud server, and checking services on the same server?
Whats your setup, whats your system? What are your monitors?

With given more detailed information it is just pointing into the dark.

**Comment by neldonado** [comment]
> > I am running into the same type of errors, I am running uptime kuma off a google cloud server running an ubuntu 20.04 LTS VM
> > It'll say UP and then bounces back with that error. Back and forth back and forth.
> 
> Can you be more specific? You running it on a cloud server, and checking services on the same server? Whats your setup, whats your system? What are your monitors?
> 
> With given more detailed information it is just pointing into the dark.

I have Kuma running on ubuntu 20.04 which is running in a VM on a Google Cloud instance. I am using Kuma to check my personal domains and subdomains.

My personal domains are redirected to NGINX Proxy Manager that is running on my unraid server at my home using duckdns for static IP. 

I host several applications with NPM and never have issues connecting to them through my subdomains. 

Running it like this I get a lot of `getaddrinfo EAI_AGAIN` errors. 

![image](https://user-images.githubusercontent.com/46225185/193441377-9bb7543d-da0f-469c-a022-e5f23b3e9c14.png)

If I run kuma on my Unraid server, where I also run NPM, I don't get these errors. 

![image](https://user-images.githubusercontent.com/46225185/193441445-e359ca14-7c5e-4caf-b610-87e56fc6e3b5.png)


I should also note I have used uptime robot for hosting all of these subdomains and have had flawless results. I am not sure if it's a configuration with how I set it up, or what.

**Comment by louislam** [comment]
Are you using the Alpine docker image or specific dns server?

**Comment by rezzorix** [comment]
From the screenshots I would suspect it could have something to do with timeouts...
@louislam is default timeout still 48000ms?  I read that somewhere but not sure.

**Comment by neldonado** [comment]
> Are you using the Alpine docker image or a special dns server?

Forgive my inexperience here, but in the ubuntu terminal I just did 

`git clone https://github.com/louislam/uptime-kuma.git`

[I followed this guide.](https://www.how2shout.com/linux/how-to-install-uptime-kuma-on-ubuntu-22-04-lts-jammy/)


However, I also downloaded the app off the home assistant community add-on store and it did the same thing.

**Comment by neldonado** [comment]
> From the screenshots I would suspect it could have something to do with timeouts...
> 
> @louislam is default timeout still 48000ms?  I read that somewhere but not sure.
> 
> 
> 
> 
> 
> 

Yeah I am not sure why my response times are so slow. I'm just using the NGINX Proxy Manager official image off of the unraid community App Store.

**Comment by rezzorix** [comment]
> Yeah I am not sure why my response times are so slow. I'm just using the NGINX Proxy Manager official image off of the unraid community App Store.

Well your homeserver most likely does have either one of this too: limited/slow connection or wrong setup network//firewall rules.
If the services at home are additionally connected via Wifi.. well it slows things down even further.

> > Are you using the Alpine docker image or a special dns server?
> 
> Forgive my inexperience here, but in the ubuntu terminal I just did
> 
> `git clone https://github.com/louislam/uptime-kuma.git`
> 
> [I followed this guide.](https://www.how2shout.com/linux/how-to-install-uptime-kuma-on-ubuntu-22-04-lts-jammy/)
> 
> However, I also downloaded the app off the home assistant community add-on store and it did the same thing.

Oh OK - so you are running the npm version - not the docker version.
You did the git clone to the google server; that's not necessary - you do not need all this files.

I suggest to run it with docker not from other sources first to narrow down whats the issue.

The easiest way to start the container and run uptime-kuma is just like written in the readme:
`docker run -d --restart=always -p 3001:3001 -v uptime-kuma:/app/data --name uptime-kuma louislam/uptime-kuma:1`

Try that and then revert if you have more questions.

**Comment by neldonado** [comment]
> > Yeah I am not sure why my response times are so slow. I'm just using the NGINX Proxy Manager official image off of the unraid community App Store.
> 
> 
> 
> Well your homeserver most likely does have either one of this too: limited/slow connection or wrong setup network//firewall rules.
> 
> If the services at home are additionally connected via Wifi.. well it slows things down even further.
> 
> 
> 
> > > Are you using the Alpine docker image or a special dns server?
> 
> > 
> 
> > Forgive my inexperience here, but in the ubuntu terminal I just did
> 
> > 
> 
> > `git clone https://github.com/louislam/uptime-kuma.git`
> 
> > 
> 
> > [I followed this guide.](https://www.how2shout.com/linux/how-to-install-uptime-kuma-on-ubuntu-22-04-lts-jammy/)
> 
> > 
> 
> > However, I also downloaded the app off the home assistant community add-on store and it did the same thing.
> 
> 
> 
> Oh OK - so you are running the npm version - not the docker version.
> 
> You did the git clone to the google server; that's not necessary - you do not need all this files.
> 
> 
> 
> I suggest to run it with docker not from other sources first to narrow down whats the issue.
> 
> 
> 
> The easiest way to start the container and run uptime-kuma is just like written in the readme:
> 
> `docker run -d --restart=always -p 3001:3001 -v uptime-kuma:/app/data --name uptime-kuma louislam/uptime-kuma:1`
> 
> 
> 
> Try that and then revert if you have more questions.
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Everything is hardwired and I have a symmetrical gigabit connection. 

I have this issue on both the NPM version as well as the docker version on unraid and home assistant.

**Comment by neldonado** [comment]
It's also weird that uptimerobot has no problem pinging it but uptime kuma does

**Comment by louislam** [comment]
It is usually related to Uptime Kuma's server side.  Your home network should be fine.

Since Uptime Kuma is working fine on your Unraid server. My guess is it may be related to Google Cloud Platform (GCP)'s Ubuntu image or GCP's pre-configured dns.

If possible, try to temporary create other VM with other Linux distribution like CentOS or Debian on GCP. Or change the system dns to 1.1.1.1.

**Comment by neldonado** [comment]
> It is usually related to Uptime Kuma's server side. Your home network should be fine.
> 
> Since Uptime Kuma is working fine on your Unraid server. My guess is it may be related to Google Cloud Platform (GCP)'s Ubuntu image or GCP's pre-configured dns.
> 
> If possible, try to temporary create other VM with other Linux distribution like CentOS or Debian on GCP. Or change the system dns to 1.1.1.1.

I am running uptime kuma from the Home Assistant add-on store on a hardwired machine on the same switch as my unraid server in my local network. The errors are not as often but still come up. 

![image](https://user-images.githubusercontent.com/46225185/193464705-adcd9eac-9403-43dc-b568-4b014dc6addc.png)

**Comment by neldonado** [comment]
> It is usually related to Uptime Kuma's server side.  Your home network should be fine.
> 
> 
> 
> Since Uptime Kuma is working fine on your Unraid server. My guess is it may be related to Google Cloud Platform (GCP)'s Ubuntu image or GCP's pre-configured dns.
> 
> 
> 
> If possible, try to temporary create other VM with other Linux distribution like CentOS or Debian on GCP. Or change the system dns to 1.1.1.1.

I figured out what my issue was and was able to solve this. I am using duck DNS as a free static IP service when I disable duck DNS and put in my public IP for my DNS records on my domain Everything works great and I have awesome response times no more errors so it looks like Duck DNS may have been throttling me.

**Comment by rezzorix** [comment]
DuckDNS as DynDNS?
That would be a simple DNS lookup... there is (should be) no throttling... most hiccups come from propagation taking time.

So basically you tell DuckDNS to change the A record for domain x every time your IP at home changes.
Depending on how often your home IP changes this can cause hiccups with a DNS lookup because your A records need to propagate first.

So some DNS servers may still have your old home IP e.g. 1.2.3.4 listed, while the record is already updated to the new one which is e.g. 11.12.13.14. 
So if the DNS server you are using is taking a bit longer to update and still shows the old record - it will result in timeouts.

**Comment by lordraiden** [comment]
Same problem here
Is and internal hosted url, I'm using AdguardHome.
If I type the url in any browser it just work, but for uptime kuma it doesn't, connections are allowed in the firewall.

![imagen](https://github.com/louislam/uptime-kuma/assets/17175395/015996b7-6894-4a83-b2ee-9eddc0ef9d59)

**Comment by jackra1n** [comment]
Im having the same issue with the same setup as @lordraiden

**Comment by louislam** [comment]
Not the same error. It is `ENOTFOUND`, which means your dns server cannot find your domain name.

Follow this and make sure you can ping inside your Uptime Kuma container.
https://github.com/louislam/uptime-kuma/wiki/Troubleshooting

**Comment by vrenjith** [comment]
I am getting the same error in the alerts that come to our slack channel. While the checks are running fine, the alerts have this text `getaddrinfo ENOTFOUND`. (this one is a real server down alert)
Any hints?
```
xxxx App
APP  [12:55](https://xxxx.slack.com/archives/xxxxx/xxxxx)
Uptime Kuma Alert
[xxxxx] [:red_circle: Down] getaddrinfo ENOTFOUND [xxxxx.com](http://xxxxx/) @channel

Uptime Kuma Alert
Message
[xxxxx] [:red_circle: Down] getaddrinfo ENOTFOUND [xxxxx](http://xxxxx) @channel
Time (Asia/Calcutta)
2023-10-19 12:55:44
```

Interestingly, the URL that appears in the alert has `http` instead of `https` even though the monitor has `https` URL

**Comment by tomiko23lol** [comment]
[ SOLVED ]
I had the same issue, all other devices in network were fine, but I was getting `ENOTFOUND `in uptime-kuma.
I connected to docker container running uptime-kuma and checked:
`cat /etc/resolv.conf`

and found this ...
`nameserver 127.0.0.53`

My DNS server that I was expecting there ( pihole ) is not there. After that I added following to docker compose file to set DNS server statically to IP I want there:
```
    dns:
      - 192.168.23.102
```
And this completely fixed my issue. 

Than I realized that my second DNS server (another pihole) I was using before as primary is running and it should not be. On this server DNS records have been missing. 

So if You are having similar problems, **check what DNS server is actually resolving requests from kuma**.

I hope this helps someone.
